### PR TITLE
Replace quality_profiles measure by a dedicated field in scanner report

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/TrackerRawInputFactory.java [hunk 1]
```diff
@@ -113,19 +115,19 @@ private DefaultIssue toIssue(LineHashSequence lineHashSeq, ScannerReport.Issue r
...
-      if (reportIssue.hasMsg()) {
+      if (isNotEmpty(reportIssue.getMsg())) {
         issue.setMessage(reportIssue.getMsg());
       }
-      if (reportIssue.hasSeverity()) {
+      if (reportIssue.getSeverity() != Severity.UNKNOWN_SEVERITY) {
         issue.setSeverity(reportIssue.getSeverity().name());
       }
-      if (reportIssue.hasGap()) {
+      if (reportIssue.getGap() != 0) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/TrackerRawInputFactory.java]: CRITICAL Equality tests should not be made with floating point values. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1244)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 2]
```diff
@@ -2554,12 +2554,16 @@
...
    */
+  @Deprecated
   public static final String QUALITY_PROFILES_KEY = "quality_profiles";
 
   /**
    * @since 4.4
+   * @deprecated since 5.5
    */
+  @Deprecated
   public static final Metric<String> QUALITY_PROFILES = new Metric.Builder(QUALITY_PROFILES_KEY, "Profiles", Metric.ValueType.DATA)
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)
MINOR Remove this use of "QUALITY_PROFILES_KEY"; it is deprecated. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 3]
```diff
@@ -2554,12 +2554,16 @@
 
   /**
    * @since 4.4
+   * @deprecated since 5.5
    */
+  @Deprecated
   public static final String QUALITY_PROFILES_KEY = "quality_profiles";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-scanner-engine/src/main/java/org/sonar/batch/issue/IssueTransformer.java [hunk 4]
```diff
@@ -93,9 +94,9 @@ public static TrackedIssue toTrackedIssue(BatchComponent component, ScannerRepor
     issue.setKey(Uuids.createFast());
     issue.setComponentKey(component.key());
     issue.setRuleKey(ruleKey);
-    issue.setGap(rawIssue.hasGap() ? rawIssue.getGap() : null);
+    issue.setGap(rawIssue.getGap() != 0 ? rawIssue.getGap() : null);
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/batch/issue/IssueTransformer.java]: CRITICAL Equality tests should not be made with floating point values. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1244)

File: sonar-scanner-engine/src/main/java/org/sonar/batch/issue/DeprecatedIssueAdapterForFilter.java [hunk 5]
```diff
@@ -88,7 +88,7 @@ public Double effortToFix() {
 
   @Override
   public Double gap() {
-    return rawIssue.hasGap() ? rawIssue.getGap() : null;
+    return rawIssue.getGap() != 0 ? rawIssue.getGap() : null;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/batch/issue/DeprecatedIssueAdapterForFilter.java]: CRITICAL Equality tests should not be made with floating point values. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1244)

File: sonar-scanner-engine/src/main/java/org/sonar/batch/issue/DefaultFilterableIssue.java [hunk 6]
```diff
@@ -61,12 +61,12 @@ public String message() {
...
   @Override
   public Integer line() {
-    return rawIssue.hasLine() ? rawIssue.getLine() : null;
+    return rawIssue.hasTextRange() ? rawIssue.getTextRange().getStartLine() : null;
   }
 
   @Override
   public Double gap() {
-    return rawIssue.hasGap() ? rawIssue.getGap() : null;
+    return rawIssue.getGap() != 0 ? rawIssue.getGap() : null;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/batch/issue/DefaultFilterableIssue.java]: CRITICAL Equality tests should not be made with floating point values. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1244)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeQProfileMeasureStep.java [hunk 7]
```diff
@@ -74,19 +77,36 @@ public QProfiles createForAny(Component component) {
...
     @Override
+    public void visitFile(Component file, Path<QProfiles> path) {
+      String languageKey = file.getFileAttributes().getLanguageKey();
+      if (languageKey == null) {
+        // No qprofile for unknown languages
+        return;
+      }
+      if (!analysisMetadataHolder.getQProfilesByLanguage().containsKey(languageKey)) {
+        // Should never occurs except if scanner report is buggy (ie list of qprofile is not consistent with file languages)
+        // TODO Should we fail fast, log?
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeQProfileMeasureStep.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeQProfileMeasureStep.java [hunk 8]
```diff
@@ -38,33 +38,36 @@
...
+    AnalysisMetadataHolder analysisMetadataHolder) {
     this.treeRootHolder = treeRootHolder;
     this.measureRepository = measureRepository;
     this.metricRepository = metricRepository;
+    this.analysisMetadataHolder = analysisMetadataHolder;
   }
 
   @Override
   public void execute() {
     Metric qProfilesMetric = metricRepository.getByKey(CoreMetrics.QUALITY_PROFILES_KEY);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeQProfileMeasureStep.java]: MINOR Remove this use of "QUALITY_PROFILES_KEY"; it is deprecated. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-scanner-engine/src/main/java/org/sonar/batch/mediumtest/TaskResult.java [hunk 9]
```diff
@@ -76,22 +77,22 @@ public void scanTaskCompleted(ProjectScanContainer container) {
...
-  private void storeReportComponents(int componentRef, String parentModuleKey, @Nullable String branch) {
+  private void storeReportComponents(int componentRef, String parentModuleKey, String branch) {
     Component component = getReportReader().readComponent(componentRef);
-    if (component.hasKey()) {
-      reportComponents.put(component.getKey() + (branch != null ? ":" + branch : ""), component);
+    if (isNotEmpty(component.getKey())) {
+      reportComponents.put(component.getKey() + (isNotEmpty(branch) ? ":" + branch : ""), component);
     } else {
-      reportComponents.put(parentModuleKey + (branch != null ? ":" + branch : "") + ":" + component.getPath(), component);
+      reportComponents.put(parentModuleKey + (isNotEmpty(branch) ? ":" + branch : "") + ":" + component.getPath(), component);
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/batch/mediumtest/TaskResult.java]: MAJOR Add parentheses to make the operator precedence explicit. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS864)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues:
* CRITICAL 4 critical
* MAJOR 1 major
* MINOR 2 minor
* INFO 3 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeQProfileMeasureStep.java [hunk 10]
```diff
@@ -74,19 +77,36 @@ public QProfiles createForAny(Component component) {
...
 
     @Override
+    public void visitFile(Component file, Path<QProfiles> path) {
+      String languageKey = file.getFileAttributes().getLanguageKey();
+      if (languageKey == null) {
+        // No qprofile for unknown languages
+        return;
+      }
+      if (!analysisMetadataHolder.getQProfilesByLanguage().containsKey(languageKey)) {
+        // Should never occurs except if scanner report is buggy (ie list of qprofile is not consistent with file languages)
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeQProfileMeasureStep.java]: I would suggest to fail in such case

File: sonar-scanner-protocol/src/main/protobuf/scanner_report.proto [hunk 11]
```diff
@@ -41,6 +41,14 @@ message Metadata {
   string branch = 3;
   int32 root_component_ref = 4;
   bool cross_project_duplication_activated = 5;
+  map<string, QProfile> qprofiles_per_language = 6;
+  
+	message QProfile {
+	  string key = 1;
+    string name = 2;
+    string language = 3;
+    int64 rulesUpdatedAt = 4;
```
Comment (julienlancelot) [inline sonar-scanner-protocol/src/main/protobuf/scanner_report.proto]: is this field really useful ?

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeQProfileMeasureStep.java [hunk 12]
```diff
@@ -74,19 +77,36 @@ public QProfiles createForAny(Component component) {
...
+        // TODO Should we fail fast, log?
+        // For now silently ignore
+        return;
+      }
+      path.parent().add(analysisMetadataHolder.getQProfilesByLanguage().get(languageKey));
+    }
+
+    @Override
+    public void visitDirectory(Component directory, org.sonar.server.computation.component.PathAwareVisitor.Path<QProfiles> path) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeQProfileMeasureStep.java]: Remove full package of org.sonar.server.computation.component.PathAwareVisitor.Path

File: server/sonar-server/src/test/java/org/sonar/server/computation/analysis/AnalysisMetadataHolderRule.java [hunk 13]
```diff
@@ -108,4 +112,15 @@ public int getRootComponentRef() {
     checkState(rootComponentRef.isInitialized(), "Root component ref has not been set");
     return rootComponentRef.getProperty();
   }
+
+  public AnalysisMetadataHolderRule setQProfilesByLanguage(Map<String, QualityProfile> qProfilesPerLanguage) {
+    this.qProfilesPerLanguage.setProperty(qProfilesPerLanguage);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/analysis/AnalysisMetadataHolderRule.java]: checkState(!this.qProfilesPerLanguage.isInitialized(), "QProfiles by language has already been set") is missing

File: server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java [hunk 14]
```diff
@@ -133,25 +134,6 @@ public void nothing_to_add_when_measure_already_exists_on_project() throws Excep
     assertThat(measureRepository.getAddedRawMeasures(PROJECT_REF)).isEmpty();
   }
 
-  @Test
-  public void nothing_to_add_when_no_qprofile_computed_on_project() throws Exception {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java]: You should keep this test by testing that measure is empty when no file in project

Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java]: You should keep this test by testing that measure is empty when no file in project

File: server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java [hunk 15]
```diff
@@ -133,25 +134,6 @@ public void nothing_to_add_when_measure_already_exists_on_project() throws Excep
...
-    treeRootHolder.setRoot(MULTI_MODULE_PROJECT);
-
-    underTest.execute();
-
-    assertThat(measureRepository.getAddedRawMeasures(PROJECT_REF)).isEmpty();
-  }
-
-  @Test
-  public void nothing_to_add_when_qprofiles_computed_on_project_are_empty() throws Exception {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java]: You should keep this test by testing that measure is empty when qprofile data from AnalysisMetadataHolderRule is empty

File: server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java [hunk 16]
```diff
@@ -76,47 +97,27 @@
...
+    QualityProfile qpPhp = createQProfile(QP_NAME_2, LANGUAGE_KEY_2);
+    analysisMetadataHolder.setQProfilesByLanguage(ImmutableMap.of(LANGUAGE_KEY_1, qpJava, LANGUAGE_KEY_2, qpPhp));
 
     underTest.execute();
 
-    assertThat(measureRepository.getAddedRawMeasures(PROJECT_REF).get(QUALITY_PROFILES_KEY)).extracting("data").containsOnly(toJson(qp));
-  }
-
-  @Test
-  public void add_quality_profile_measure_from_multiple_modules() throws Exception {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java]: This use case is not needed anymore ?

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 17]
```diff
@@ -394,21 +394,21 @@
   @Deprecated
   public static final Metric<String> CLASS_COMPLEXITY_DISTRIBUTION = new Metric.Builder(CLASS_COMPLEXITY_DISTRIBUTION_KEY, "Classes distribution /complexity",
     Metric.ValueType.DISTRIB)
-    .setDescription("Classes distribution /complexity")
-    .setDirection(Metric.DIRECTION_NONE)
-    .setQualitative(true)
-    .setDomain(DOMAIN_COMPLEXITY)
-    .setHidden(true)
-    .create();
+      .setDescription("Classes distribution /complexity")
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: Your configuration seems to be not good, all these spaces should not be added

**Comment by julienlancelot** [comment]
@henryju Please apply my feedbacks

File: server/sonar-server/src/test/java/org/sonar/server/computation/analysis/AnalysisMetadataHolderRule.java (hunk 13 above)
Comment (henryju) [inline server/sonar-server/src/test/java/org/sonar/server/computation/analysis/AnalysisMetadataHolderRule.java]: Not done on other setter...

File: server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java (hunk 16 above)
Comment (henryju) [inline server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeQProfileMeasureStepTest.java]: Previsoulsy there was a special case for modules (need to merge QP measure) but this is no more the case.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MINOR 2 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MINOR 2 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Pull request merged**